From 13458c7cd934852d9d17948448b80d90ba485476 Mon Sep 17 00:00:00 2001
From: Robert Kuban <rkuban@qti.qualcomm.com>
Date: Thu, 28 Aug 2025 15:53:40 +0200
Subject: [PATCH] qcom_scm-coqoshv: Initialize SCM before entering atomic
 kernel context

`coqoshv_call_qcpe` initializes the channel to QCPE on the first call.
Unfortunately the first call also queries the SMC convention inside
an atomic kernel context, and the functions needed for initialization
may block.

This can be detected if the kernel is configured with
CONFIG_DEBUG_ATOMIC_SLEEP && CONFIG_PANIC_ON_SCHED_BUG.

Initialize the channel before entering the atomic kernel context.

Signed-off-by: Robert Kuban <rkuban@qti.qualcomm.com>
---
 drivers/firmware/qcom_scm-coqoshv.c | 8 +++-----
 drivers/firmware/qcom_scm-coqoshv.h | 2 ++
 drivers/firmware/qcom_scm-smc.c     | 6 +++++-
 3 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/drivers/firmware/qcom_scm-coqoshv.c b/drivers/firmware/qcom_scm-coqoshv.c
index 61d7a64a2..3df4c1faa 100644
--- a/drivers/firmware/qcom_scm-coqoshv.c
+++ b/drivers/firmware/qcom_scm-coqoshv.c
@@ -90,8 +90,6 @@ static unsigned long coqoshv_mailbox_alloc = 0;
 static unsigned long coqoshv_mailbox_max = 0;
 static unsigned long coqoshv_mailbox_notified = ~0ul;
 
-static int coqoshv_init_qcpe(void);
-
 static void coqoshv_notify_mailboxes()
 {
 	__u32 shmem_res_id, local_res_id;
@@ -203,8 +201,7 @@ void coqoshv_call_qcpe(const struct coqoshv_smc_mailbox *smc,
 #endif
 	coqoshv_wait_sem(&coqoshv_buffer_semaphore, atomic);
 
-	if (unlikely(!coqoshv_initialized))
-		coqoshv_init_qcpe();
+	BUG_ON(!coqoshv_initialized);
 
 	mbox_id = coqoshv_acquire_mailbox();
 	msg_id = READ_ONCE(coqoshv_buffer.mheader[mbox_id].req_id) + 1;
@@ -439,8 +436,9 @@ static int coqoshv_platform_init(void)
 	return ret;
 }
 
-static int coqoshv_init_qcpe(void)
+int coqoshv_init_qcpe(void)
 {
+	BUG_ON(coqoshv_initialized);
 	if (coqoshv_platform_init())
 		panic("Could not initialize.\n");
 #if IS_ENABLED(CONFIG_QCOM_SCM_SYS_LISTENER_HOS)
diff --git a/drivers/firmware/qcom_scm-coqoshv.h b/drivers/firmware/qcom_scm-coqoshv.h
index a124ac1bc..6447d747b 100644
--- a/drivers/firmware/qcom_scm-coqoshv.h
+++ b/drivers/firmware/qcom_scm-coqoshv.h
@@ -13,3 +13,5 @@ struct coqoshv_smc_mailbox {
 
 void coqoshv_call_qcpe(const struct coqoshv_smc_mailbox *smc,
 		       struct arm_smccc_res *res, const bool atomic);
+
+int coqoshv_init_qcpe(void);
diff --git a/drivers/firmware/qcom_scm-smc.c b/drivers/firmware/qcom_scm-smc.c
index c443862d9..81da1f137 100644
--- a/drivers/firmware/qcom_scm-smc.c
+++ b/drivers/firmware/qcom_scm-smc.c
@@ -1134,8 +1134,12 @@ static void __query_convention(void)
 
 static inline enum qcom_smc_convention __get_convention(void)
 {
-	if (unlikely(!has_queried))
+	if (unlikely(!has_queried)) {
+#if IS_ENABLED(CONFIG_QCOM_SCM_QCPE_HOS)
+		coqoshv_init_qcpe();
+#endif
 		__query_convention();
+	}
 	return qcom_smc_convention;
 }
 
-- 
2.49.0

